### PR TITLE
[#218] Change redirects default behaviour

### LINKED ISSUE #218: [BUG] Investigate 503
# Description

In some of our gitlab projects, we have links such as `https://gitlab.com/morley-framework/morley-infra/-/issues/new`.

For some reason, when xrefcheck attempts to verify these links, it gets a 503 Service Temporarily Unavailable, which causes xrefcheck to fail. As a result, our gitlab projects ignore those links with a `xrefcheck: ignore link` annotation.

However, if you manually verify it using curl or using a browser, you get a 302 redirection (which would make xrefcheck pass).

* Let's investigate the reason for this discrepancy.
* If there's anything that can be doe in xrefcheck to fix this, let's do it.
* If we manage to fix it, let's remove the `xrefcheck: ignore annotation` from those gitlab projects. These are the ones I'm aware of:
    - [ ] https://gitlab.com/morley-framework/morley-debugger
    - [ ] https://gitlab.com/morley-framework/morley-infra
    - [ ] https://gitlab.com/morley-framework/indigo/-/issues

# To Reproduce

```
$ echo '[a](https://gitlab.com/morley-framework/morley-infra/-/issues/new)' > a.md
$ git add a.md
$ xrefcheck
```

```
Configuration file not found, using default config for GitHub repositories

=== Invalid references found ===                                                                                                                    

  ➥  In file a.md
     bad reference (external) at src:7:1-66:
       - text: "a"
       - link: https://gitlab.com/morley-framework/morley-infra/-/issues/new
       - anchor: -
     
     ⛂  Resource unavailable (503 Service Temporarily Unavailable)
     
Invalid references dumped, 1 in total.
```

However, if we check the same link via curl, it replies with a 302 redirect.

```
$ curl -I https://gitlab.com/morley-framework/morley-infra/-/issues/new
HTTP/2 302 
...
location: https://gitlab.com/users/sign_in
...
```

Similarly, opening the link in the browser (while not authenticated into gitlab.com) gets us redirected to `https://gitlab.com/users/sign_in`.

# Expected behavior

Expected the link to reply with a 302 and for xrefcheck to pass.

# Environment

* xrefcheck v0.2.2

### PR DESCRIPTION
**Pull request by aeqz**
## Description

Problem: Xrefcheck currently always follows redirect links.

Solution: We are changing its default behaviour regarding redirect links to fail and report permanent redirects, and to pass for temporary redirects. Further PRs will allow the user to configure other policies.

## Related issue(s)

Fixes #218
Relates #25

## :white_check_mark: Checklist for your Pull Request

Ideally a PR has all of the checkmarks set.

If something in this list is irrelevant to your PR, you should still set this
checkmark indicating that you are sure it is dealt with (be that by irrelevance).

#### Related changes (conditional)

- Tests
  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](https://github.com/serokell/xrefcheck/tree/master/README.md)
    - Haddock

- Public contracts
  - [x] Any modifications of public contracts comply with the [Evolution
  of Public Contracts](https://www.notion.so/serokell/Evolution-of-Public-Contracts-2a3bf7971abe4806a24f63c84e7076c5) policy.
  - [x] I added an entry to the [changelog](https://github.com/serokell/xrefcheck/tree/master/CHANGES.md) if my changes are visible to the users
        and
  - [x] provided a migration guide for breaking changes if possible

#### Stylistic guide (mandatory)

- [x] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Where-and-how-to-commit-your-work-58f8973a4b3142c8abbd2e6fd5b3a08e).
- [x] My code complies with the [style guide](https://github.com/serokell/style/blob/master/haskell.md).

### CODE REVIEW

File: tests/golden/check-autolinks/check-autolinks.bats
Comment (aeqz) [inline tests/golden/check-autolinks/check-autolinks.bats:39]: Will this become a problem? HTTP requests that redirect to their respective HTTPS version.

Comment (Martoon-00) [reply tests/golden/check-autolinks/check-autolinks.bats:39]: I think this is fine, IMO the user should provide the direct `https` link in such case so we fit the general rule.

File: src/Xrefcheck/Verify.hs
Comment (Martoon-00) [inline src/Xrefcheck/Verify.hs:727]: Minor nitpick: it is less boilerplate-y to use `\case`, but feel free to leave the current version if you like this syntax. Or, maybe even shorter: use `elem`.

File: src/Xrefcheck/Verify.hs
Comment (Martoon-00) [inline src/Xrefcheck/Verify.hs:750]: Extra space?

File: tests/golden/check-autolinks/check-autolinks.bats
Comment (Martoon-00) [reply tests/golden/check-autolinks/check-autolinks.bats:39]: Even if `http` was an auto-inserted default. AFAIU if the user dislikes this behaviour, it will be configurable too.

Maybe when describing configurable redirects later in `README`, we can provide an example with suppresses errors about `http -> https`.

Comment (Martoon-00) [reply tests/golden/check-autolinks/check-autolinks.bats:39]: So we even immediately have one test for this, neat :smile:

File: src/Xrefcheck/Verify.hs
Comment (Martoon-00) [inline src/Xrefcheck/Verify.hs:251]: Perhaps I would put the links on separate lines, for two reasons:
1. They might be large;
2. This way they will be easier to compare visually. For instance, in case of `http` -> `https` suggestion, if the link is large then the user may struggle to see what is the difference.

**Review by Martoon-00 — COMMENTED** [review]
Looks good, thank you :+1: 

Leaving a couple of insignificant concerns, and a medium one.

**Comment by Martoon-00** [comment]
Oh, and one more thing: make sure to fit commit descriptions into 80 characters.

File: src/Xrefcheck/Verify.hs
Comment (aeqz) [reply src/Xrefcheck/Verify.hs:251]: Right, this "from `http` to `https`" example is a clear case in which having the links reported in separate lines can be useful.

File: tests/Test/Xrefcheck/UtilRequests.hs
Comment (aeqz) [inline tests/Test/Xrefcheck/UtilRequests.hs:6]: I have refactored some functions from `Test.Xrefcheck.TooManyRequestsSpec` in order to reuse them in `Test.Xrefcheck.RedirectRequestsSpec`.

File: tests/Test/Xrefcheck/RedirectRequestsSpec.hs
Comment (Martoon-00) [inline tests/Test/Xrefcheck/RedirectRequestsSpec.hs:84]: This is only my opinion, but I believe there is no reason to provide that high coverage, since e.g. handling of `302` and `303` in the code is exactly the same. E.g. having 302 w/ loc, 302 w/o loc, 307 w/ loc would be sufficient to cover the temporary redirects.

Using formal terms, we don't usually practice automated "black box" testing when the implementation is only one and concrete.

What do you think?

File: tests/Test/Xrefcheck/TooManyRequestsSpec.hs
Comment (Martoon-00) [inline tests/Test/Xrefcheck/TooManyRequestsSpec.hs:160]: :+1:

**Review by Martoon-00 — COMMENTED** [review]
Nice to see tests added :+1:   I think you can add a tick to the PR description.

Leaving one more comment, would like to know your opinion.

File: tests/Test/Xrefcheck/RedirectRequestsSpec.hs
Comment (YuriRomanowski) [reply tests/Test/Xrefcheck/RedirectRequestsSpec.hs:84]: I think that it's OK to check all the possible variants. But it seems that way these tests are written is a bit verbose. We could for example make a separate function which tests two options (location present/absent) for a given code, and pass to this function as a parameter whether these tests should succeed or not. I believe that code will get much shorter this way.

Comment (Martoon-00) [reply tests/Test/Xrefcheck/RedirectRequestsSpec.hs:84]: If test cases can be turned into one- or two-liners, that approach sounds good.

Comment (Martoon-00) [reply tests/Test/Xrefcheck/RedirectRequestsSpec.hs:26]: Looks nice :+1:

**Review by Martoon-00 — APPROVED** [review]

**Comment by Martoon-00** [comment]
@aeqz The comment of @YuriRomanowski seems to be also addressed, so in theory, you can merge this PR.

But indeed it is better to wait for his approval or a "feel free to merge without my approval" message.

**Comment by Martoon-00** [comment]
Don't forget to squash your commits, as the 2nd and 3rd seem to be just additions to the 1st commit.

**Review by YuriRomanowski — APPROVED** [review]

**Pull request merged**
